### ISSUE TITLE
Cannot change columns in log-file after it is created

### ISSUE THREAD

**Issue by 0xxon** [zeek/zeek#1647]
In #1634, it was mentioned that it would be nice to make an option that changes the log-fields configurable during run-time.

Testing this showed that we currently cannot modify the columns of a log-file after it has been written to. We potentially should update our writers to just output a new header (for the ASCII one), or to be able to otherwise deal with this situation.

Example script:

```
module SSH;

export {
        redef enum Log::ID += { LOG };

        type Log: record {
                t: time;
                id: conn_id; # Will be rolled out into individual columns.
                status: string &optional;
                country: string &default="unknown";
        } &log;
}

event zeek_init()
{
        Log::create_stream(SSH::LOG, [$columns=Log]);

        Log::remove_default_filter(SSH::LOG);
        Log::add_filter(SSH::LOG, [$name="default", $exclude=set("t", "id.orig_h")]);

        local cid = [$orig_h=1.2.3.4, $orig_p=1234/tcp, $resp_h=2.3.4.5, $resp_p=80/tcp];

        Log::write(SSH::LOG, [$t=network_time(), $id=cid, $status="success"]);
        Log::write(SSH::LOG, [$t=network_time(), $id=cid, $status="failure", $country="US"]);
        Log::write(SSH::LOG, [$t=network_time(), $id=cid, $status="failure", $country="UK"]);

        Log::remove_default_filter(SSH::LOG);
        Log::add_filter(SSH::LOG, [$name="default", $exclude=set("t")]);

        Log::write(SSH::LOG, [$t=network_time(), $id=cid, $status="success", $country="BR"]);
        Log::write(SSH::LOG, [$t=network_time(), $id=cid, $status="failure", $country="MX"]);

}
```

Running with zeek -b results in:

```
warning: WriterFrontend ssh/Log::WRITER_ASCII expected 5 fields in write, got 6. Skipping line.
warning: WriterFrontend ssh/Log::WRITER_ASCII expected 5 fields in write, got 6. Skipping line.
```

**Comment by ckreibich** [comment]
I have a conceptual question here — is this something we want to support? Won't it often confuse data ingesters if the data format changes at runtime? I suppose we could argue that's the ingesters' problem, and we can/should support it. But I could see it create confusion.

**Comment by mavam** [comment]
Coming from the perspective of an ingester, I can tell you the answer is it depends on whether there's a static type mapping downstream. If you're just inhaling the JSON and don't care too much about typing, then this might be fine. If you're converting Zeek data into structured data, then this may matter.

For example, [VAST](https://vast.io) can handle the dynamic nature of Zeek TSV just fine, but we'd need an out-of-band schema declaration for the JSON data to apply deeper semantics than just strings and numbers. It's probably the same with Corelight's ECS mappings.

**Comment by rsmmr** [comment]
I'd also question if this is a feature we want to support. I'd also add that having new headers in the middle of a Zeek log isn't something that ingesters might expect to encounter to begin with (independent of whether fields are changing). I don't recall if we ever suggested that that's possible, but even if we did, people might just not be prepared for it. 

That still leaves the question how to handle the situation, the current behavior isn't great obviously. Maybe the stream configuration should just be frozen after the first write?

**Comment by mavam** [comment]
> I'd also add that having new headers in the middle of a Zeek log isn't something that ingesters might expect to encounter to begin with

Agreed, it's safe to assume that's not the case.

> Maybe the stream configuration should just be frozen after the first write?

Sounds prudent to me when a fixed schema is the common denominator. Attempting to make changes after the stream is "live" (i.e., first record written) should result an error.

**Comment by ckreibich** [comment]
I've looked at this a bit ... the script-level framework code is all about filters, but the error message actually comes from the ASCII writer that the framework maintains in the core, associated with the `SSH::LOG` stream. The writer survives the replacement of the default filter, and the new filter's columns no longer match its expectations. The writer can't just ignore the mismatch because it was originally created to write a certain column layout.

We could expand `Manager::AddFilter()` to use some of the writer-locating code that's currently in `Manager::Write()`. When that code locates an existing writer, and that writer is incompatible with the new filter's effective columns, the filter addition should fail. Thankfully it already indicates success, so we'd just add another failure scenario.

What do you guys think?

**Comment by awelzel** [comment]
> When that code locates an existing writer, and that writer is incompatible with the new filter's effective columns, the filter addition should fail. Thankfully it already indicates success, so we'd just add another failure scenario.
> What do you guys think?

I might even suggest to hard exit. Minimally a very loud `reporter->Error()` and/or output on stderr in addition. Having `Log::add_filter` fail probably goes undetected until someone realizes the logs don't look what they expect.  Even the reporter.log messages may go unnoticed for sometime unless you're very diligent (@sowmyaramapatruni ran into this quite recently, actually).

**Comment by ckreibich** [comment]
Yeah agree, though exiting strikes me as quite strong ... I just looked over the existing failure reasons for `Log::add_filter()`. There are quite a few, mostly around type-checking of the provided filter record, and log stream existence. The existing uses of `reporter->FatalError()` are mostly due to internal problems, not usually something the user does in scriptland. So I vote for `reporter->Error()` for this one.

Also it's true that people aren't currently likely to check the `Log::add_filter()` result, even though the docs clearly state that it can fail. So it'd be good if we could flag that ... seems like a nice one for future linting / `-u` additions.

**Comment by stevesmoot** [comment]
I haven’t thought deeply about it but is there A serious use case for after
zeek_init ? For generating a schema it would be nice to run at a low
priority of zeek_init.

S
